design/design.md

```diff
@@ -0,0 +1,232 @@
+# Design for Streamly Coreutils package
+
+* cp
+
+  ```
+      data CpOptions = CpOptions {
+         attributesOnly :: Bool,
+         interactive :: Bool,
+         parents :: Bool,
+         symbolicLink :: Bool,
+         noTargetDir :: Bool,
+         verbose :: Bool,
+         noTargetDir :: Bool
+         ...
+      }
+
+      cpFile :: CpOptions -> SomeBase File -> SomeBase Dir -> SomeBase File -> IO ()
+
+  ```
+
+  1. Similar to srcFile -> destDir -> destFile
+  2. Copies the source file to a dest directory with dest file name
+  3. If dest file name is empty, then it becomes File -> Dir case (with same file name as source)
+
+  4. For File -> File within same directory, current directory should be passed as destDir argument
+   Else we can also say that an empty Dest Dir argument is equivalent to copying in the same directory
+
+  ```
+      cpFiletoDir :: (IsStream t, Monad m) => CpOptions -> t m (SomeBase File) -> SomeBase Dir -> IO ()
+         uses cpFile
+         (stream of input files so that copying can be done concurrently with use of appropriate streams)
+         or map can be used as well
+
+      cpDirToDir :: (IsStream t, Monad m) => CpOptions -> SomeBase Dir -> SomeBase Dir -> IO ()
+         should create a stream of files (t m (SomeBase File)) and call cpFileToDir
+
+  ```
+* echo
+
+  ```
+      data EchoOptions = CpOptions {
+         trailingLine :: Bool,
+         interpretBackSlash :: Bool,
+         ...
+      }
+
+      echo :: (IsStream t, Monad m) => EchoOptions -> t m Char -> SomeBase File -> IO ()
+
+  ```
+
+   1. for stdout, second arg should be "/dev/stdout"
```

Comment:
Oh, I Just saw this. My bad! Exploit the type system as much as you can. `"/dev/stdout"` is not platform agnostic.

Reply:
Better to use a `Handle` instead.

Reply:
Okay then we should remove `stdout`

Reply:
What do you mean by "remove `stdout`"?

Reply:
Since you said it's not platform agnostic, we should not use it. But I don't know its equivalents in Mac/Windows.